// @storybook/core/dist/server/config/webpack.config.default.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createDefaultWebpackConfig = createDefaultWebpackConfig;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _autoprefixer = _interopRequireDefault(require("autoprefixer"));

function createDefaultWebpackConfig(storybookBaseConfig) {
  return (0, _objectSpread2.default)({}, storybookBaseConfig, {
    module: (0, _objectSpread2.default)({}, storybookBaseConfig.module, {
      rules: (0, _toConsumableArray2.default)(storybookBaseConfig.module.rules).concat([{
        test: /\.css$/,
        use: [require.resolve('style-loader'), {
          loader: require.resolve('css-loader'),
          options: {
            importLoaders: 1
          }
        }, {
          loader: require.resolve('postcss-loader'),
          options: {
            ident: 'postcss',
            // https://webpack.js.org/guides/migrating/#complex-options
            postcss: {},
            plugins: function plugins() {
              return [require('postcss-flexbugs-fixes'), // eslint-disable-line
              (0, _autoprefixer.default)({
                flexbox: 'no-2009'
              })];
            }
          }
        }]
      }, {
        test: /\.(ico|jpg|jpeg|png|gif|eot|otf|webp|ttf|woff|woff2)(\?.*)?$/,
        loader: require.resolve('file-loader'),
        query: {
          name: 'static/media/[name].[hash:8].[ext]'
        }
      }, {
        test: /\.svg$/,
        loader: require.resolve('svg-url-loader')
      }, {
        test: /\.(mp4|webm|wav|mp3|m4a|aac|oga)(\?.*)?$/,
        loader: require.resolve('url-loader'),
        query: {
          limit: 10000,
          name: 'static/media/[name].[hash:8].[ext]'
        }
      }])
    })
  });
}
