// @storybook/core/dist/server/mergeConfigs.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

function plugins(_ref, _ref2) {
  var _ref$plugins = _ref.plugins,
      defaultPlugins = _ref$plugins === void 0 ? [] : _ref$plugins;
  var _ref2$plugins = _ref2.plugins,
      customPlugins = _ref2$plugins === void 0 ? [] : _ref2$plugins;
  return (0, _toConsumableArray2.default)(defaultPlugins).concat((0, _toConsumableArray2.default)(customPlugins));
}

function rules(_ref3, _ref4) {
  var _ref3$rules = _ref3.rules,
      defaultRules = _ref3$rules === void 0 ? [] : _ref3$rules;
  var _ref4$rules = _ref4.rules,
      customRules = _ref4$rules === void 0 ? [] : _ref4$rules;
  return (0, _toConsumableArray2.default)(defaultRules).concat((0, _toConsumableArray2.default)(customRules));
}

function extensions(_ref5, _ref6) {
  var _ref5$extensions = _ref5.extensions,
      defaultExtensions = _ref5$extensions === void 0 ? [] : _ref5$extensions;
  var _ref6$extensions = _ref6.extensions,
      customExtensions = _ref6$extensions === void 0 ? [] : _ref6$extensions;
  return (0, _toConsumableArray2.default)(defaultExtensions).concat((0, _toConsumableArray2.default)(customExtensions));
}

function alias(_ref7, _ref8) {
  var _ref7$alias = _ref7.alias,
      defaultAlias = _ref7$alias === void 0 ? {} : _ref7$alias;
  var _ref8$alias = _ref8.alias,
      customAlias = _ref8$alias === void 0 ? {} : _ref8$alias;
  return (0, _objectSpread2.default)({}, defaultAlias, customAlias);
}

function _module(_ref9, _ref10) {
  var _ref9$module = _ref9.module,
      defaultModule = _ref9$module === void 0 ? {} : _ref9$module;
  var _ref10$module = _ref10.module,
      customModule = _ref10$module === void 0 ? {} : _ref10$module;
  return (0, _objectSpread2.default)({}, defaultModule, customModule, {
    rules: rules(defaultModule, customModule)
  });
}

function resolve(_ref11, _ref12) {
  var _ref11$resolve = _ref11.resolve,
      defaultResolve = _ref11$resolve === void 0 ? {} : _ref11$resolve;
  var _ref12$resolve = _ref12.resolve,
      customResolve = _ref12$resolve === void 0 ? {} : _ref12$resolve;
  return (0, _objectSpread2.default)({}, defaultResolve, customResolve, {
    alias: alias(defaultResolve, customResolve),
    extensions: extensions(defaultResolve, customResolve)
  });
}

function mergeConfigs(config, customConfig) {
  return (0, _objectSpread2.default)({}, customConfig, config, {
    devtool: customConfig.devtool || config.devtool,
    plugins: plugins(config, customConfig),
    module: _module(config, customConfig),
    resolve: resolve(config, customConfig)
  });
}

var _default = mergeConfigs;
exports.default = _default;
