// @storybook/core/dist/server/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseList = parseList;
exports.getEnvConfig = getEnvConfig;
exports.getMiddleware = getMiddleware;
exports.getPreviewBodyHtml = getPreviewBodyHtml;
exports.getPreviewHeadHtml = getPreviewHeadHtml;
exports.getManagerHeadHtml = getManagerHeadHtml;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _path = _interopRequireDefault(require("path"));

var _fs = _interopRequireDefault(require("fs"));

function parseList(str) {
  return str.split(',');
}

function getEnvConfig(program, configEnv) {
  Object.keys(configEnv).forEach(function (fieldName) {
    var envVarName = configEnv[fieldName];
    var envVarValue = process.env[envVarName];

    if (envVarValue) {
      program[fieldName] = envVarValue; // eslint-disable-line
    }
  });
}

function getMiddleware(configDir) {
  var middlewarePath = _path.default.resolve(configDir, 'middleware.js');

  if (_fs.default.existsSync(middlewarePath)) {
    var middlewareModule = require(middlewarePath); // eslint-disable-line


    if (middlewareModule.__esModule) {
      // eslint-disable-line
      middlewareModule = middlewareModule.default;
    }

    return middlewareModule;
  }

  return function () {};
}

var interpolate = function interpolate(string) {
  var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return Object.entries(data).reduce(function (acc, _ref) {
    var _ref2 = (0, _slicedToArray2.default)(_ref, 2),
        k = _ref2[0],
        v = _ref2[1];

    return acc.replace("%".concat(k, "%"), v);
  }, string);
};

function getPreviewBodyHtml() {
  return _fs.default.readFileSync(_path.default.resolve(__dirname, 'templates/base-preview-body.html'), 'utf8');
}

function getPreviewHeadHtml(configDirPath, interpolations) {
  var base = _fs.default.readFileSync(_path.default.resolve(__dirname, 'templates/base-preview-head.html'), 'utf8');

  var headHtmlPath = _path.default.resolve(configDirPath, 'preview-head.html');

  var result = base;

  if (_fs.default.existsSync(headHtmlPath)) {
    result += _fs.default.readFileSync(headHtmlPath, 'utf8');
  }

  return interpolate(result, interpolations);
}

function getManagerHeadHtml(configDirPath, interpolations) {
  var base = _fs.default.readFileSync(_path.default.resolve(__dirname, 'templates/base-manager-head.html'), 'utf8');

  var scriptPath = _path.default.resolve(configDirPath, 'manager-head.html');

  var result = base;

  if (_fs.default.existsSync(scriptPath)) {
    result += _fs.default.readFileSync(scriptPath, 'utf8');
  }

  return interpolate(result, interpolations);
}
